src/compute/api.rs

```diff
@@ -53,151 +56,253 @@ pub fn create_keypair(session: &Session, request: KeyPairCreate) -> Result<KeyPa
+    let root: CreatedServerRoot = session
+        .post(COMPUTE, &["servers"])
+        .json(&body)
+        .fetch()
+        .await?;
     trace!("Requested creation of server {:?}", root.server);
     Ok(root.server)
 }
 
 /// Delete a key pair.
-pub fn delete_keypair<S: AsRef<str>>(session: &Session, name: S) -> Result<()> {
+pub async fn delete_keypair<S: AsRef<str>>(session: &Session, name: S) -> Result<()> {
     debug!("Deleting key pair {}", name.as_ref());
-    let _ = session.delete(COMPUTE, &["os-keypairs", name.as_ref()], None)?;
+    let _ = session
+        .delete(COMPUTE, &["os-keypairs", name.as_ref()])
+        .send()
+        .await?;
     debug!("Key pair {} was deleted", name.as_ref());
     Ok(())
 }
 
 /// Delete a server.
-pub fn delete_server<S: AsRef<str>>(session: &Session, id: S) -> Result<()> {
+pub async fn delete_server<S: AsRef<str>>(session: &Session, id: S) -> Result<()> {
     trace!("Deleting server {}", id.as_ref());
-    let _ = session.delete(COMPUTE, &["servers", id.as_ref()], None)?;
+    let _ = session
+        .delete(COMPUTE, &["servers", id.as_ref()])
+        .send()
+        .await?;
     debug!("Successfully requested deletion of server {}", id.as_ref());
     Ok(())
 }
 
 /// Get a flavor by its ID.
-pub fn get_extra_specs_by_flavor_id<S: AsRef<str>>(
+pub async fn get_extra_specs_by_flavor_id<S: AsRef<str>>(
     session: &Session,
     id: S,
 ) -> Result<HashMap<String, String>> {
     trace!("Get compute extra specs by ID {}", id.as_ref());
-    let root: ExtraSpecsRoot =
-        session.get_json(COMPUTE, &["flavors", id.as_ref(), "os-extra_specs"], None)?;
+    let root: ExtraSpecsRoot = session
+        .get_json(COMPUTE, &["flavors", id.as_ref(), "os-extra_specs"])
+        .await?;
     trace!("Received {:?}", root.extra_specs);
     Ok(root.extra_specs)
 }
 
 /// Get a flavor.
-pub fn get_flavor<S: AsRef<str>>(session: &Session, id_or_name: S) -> Result<Flavor> {
+pub async fn get_flavor<S: AsRef<str>>(session: &Session, id_or_name: S) -> Result<Flavor> {
     let s = id_or_name.as_ref();
-    get_flavor_by_id(session, s).if_not_found_then(|| get_flavor_by_name(session, s))
+    match get_flavor_by_id(session, s).await {
+        Ok(value) => Ok(value),
+        Err(err) if err.kind() == ErrorKind::ResourceNotFound => {
+            get_flavor_by_name(session, s).await
+        }
+        Err(err) => Err(err),
+    }
 }
 
 /// Get a flavor by its ID.
-pub fn get_flavor_by_id<S: AsRef<str>>(session: &Session, id: S) -> Result<Flavor> {
+pub async fn get_flavor_by_id<S: AsRef<str>>(session: &Session, id: S) -> Result<Flavor> {
     trace!("Get compute flavor by ID {}", id.as_ref());
-    let version = flavor_api_version(session)?;
-    let root: FlavorRoot = session.get_json(COMPUTE, &["flavors", id.as_ref()], version)?;
+    let version = flavor_api_version(session).await?.ok_or_else(|| {
+        Error::new(
```

Comment:
That's right, I don't remember exactly why I did that.  
Maybe it was because the `api_version` and `set_api_version` methods on `ServiceRequestBuilder` don't take an `Option`, but this doesn't excuse that my change was indeed incorrect.  
All occurrences of this problem should now be fixed.